src/torrent.cpp

```diff
@@ -1138,6 +1138,50 @@ bool is_downloading_state(int const st)
 		if (m_peer_list) m_peer_list->clear();
 	}
 
+	void torrent::set_sequential_range(piece_index_t first_piece, piece_index_t last_piece)
+	{
+		if (!has_picker()) {
+			if (!valid_metadata() || !m_connections_initialized) return;
+			else if (!m_have_all
+				|| settings().get_int(settings_pack::suggest_mode)
+				== settings_pack::suggest_read_cache)
+				need_picker();
+			else return;
+		}
+		m_sequential_download = true;
+		m_picker->set_sequential_range(first_piece, last_piece);
+	}
+
+	void torrent::set_sequential_start(piece_index_t first_piece)
+	{
+		if (!has_picker()) {
+			if (!valid_metadata() || !m_connections_initialized) return;
+			else if (!m_have_all
+				|| settings().get_int(settings_pack::suggest_mode)
+				== settings_pack::suggest_read_cache)
+				need_picker();
+			else return;
+		}
+		// m_torrent_file->files() is file_storage
```

Comment:
I think `first_piece` needs to be validated here too, to ensure it's within the valid range of pieces, and return as a no-op if it isn't

Reply:
this comment should be removed.
You validate `first_piece` in the `torrent_handle`, but you don't validate that it's < num_pieces. I think you need to do that here, that you have access to the metadata, and you know how many pieces there are.